### PR TITLE
[`OWL-VIT`] Added sdpa attention

### LINKED ISSUE #28103: OWL-VIT Vision Foundation Model deployment in the edge cases - Need SDPA support for OWL-ViT Model optimization for Edge Deployment
### Feature request

Hi Team,
I am working with  OWL-ViT Size model which has around 611 MB size ( https://huggingface.co/google/owlvit-base-patch16).
I want to optimize this model and like to deploy in the edge device for object detection.

Come to know from the group torch.scaled_dot_product_attention can be used for model optimization. 

I need your feedback comments how optimally we can reduce the memory size so that we can deploy in the edge device.

waiting for your response.

with thanks

### Motivation

It will help to deploy the models in edge so that more applications we can use it.

### Your contribution

Like to know your feedback comments.

### PR DESCRIPTION
**Pull request by nileshkokane01**
# What does this PR do?

This PR add sdpa attention for OWL-ViT.

Fixes #28103

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@NielsRogge  @younesbelkada
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by nileshkokane01** [comment]
I have added an initial draft for sdpa attention, but I guess,  more changes are needed as the OWL-ViT is a bit different compared to llama or Mistral.

Can you please point me out  a similar model close to OWL-ViT or  rather let me know the additional changes required in the file. Also, casual_attention_mask is not handled correctly - have no clue how to handle. Additionally, a corresponding test case is also necessary.

**Comment by ArthurZucker** [comment]
fyi @younesbelkada

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 1]
```diff
@@ -462,7 +524,7 @@ class OwlViTEncoderLayer(nn.Module):
     def __init__(self, config: OwlViTConfig):
```
Comment (younesbelkada) [inline src/transformers/models/owlvit/modeling_owlvit.py:524]:
```suggestion
    # Ignore Copy
    def __init__(self, config: OwlViTConfig):
```
This should fix the failing quality tests

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 2]
```diff
@@ -24,7 +24,10 @@
 from torch import Tensor, nn
 
 from ...activations import ACT2FN
-from ...modeling_attn_mask_utils import _create_4d_causal_attention_mask, _prepare_4d_attention_mask
+from ...modeling_attn_mask_utils import (
+    _create_4d_causal_attention_mask,
+    _prepare_4d_attention_mask,
+)
```
Comment (younesbelkada) [inline src/transformers/models/owlvit/modeling_owlvit.py:30]:
```suggestion
from ...modeling_attn_mask_utils import  _create_4d_causal_attention_mask, 
 _prepare_4d_attention_mask
```

**Review by younesbelkada — COMMENTED** [review]
Good work, thanks ! I left few nits, and a suggestion to fix the failing CI
Could you follow the same logic as here: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L1032-L1045 for preparing the attention mask for SDPA ? 
Also similarly as Llama, could you add `_supports_sdpa=True` in `OwlViTPreTrainedModel` ? That way the tests would be triggered (e.g.: https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/src/transformers/models/llama/modeling_llama.py#L854)

**Review by younesbelkada — COMMENTED** [review]
Hi @nileshkokane01 
Thanks a lot for your hardwork ! It looks much cleaner ! We're almost there - it seems some tests are failing: 
```bash
FAILED tests/models/owlvit/test_modeling_owlvit.py::OwlViTTextModelTest::test_model_outputs_equivalence - TypeError: _prepare_4d_causal_attention_mask_for_sdpa() missing 1 required positional argument: 'past_key_values_length'
```
Are you able to repro these failures locally? I think the fix should be to hardcode `past_key_values_length` to 0 during the call of that method as OwlViT is not a generative text model, hence does not use caching mechanism

**Comment by nileshkokane01** [comment]
@younesbelkada ,
I'm trying to solve the errors. I'll let you know when its all ready or if I need any assistance.

**Comment by nileshkokane01** [comment]
@younesbelkada  ,

I get the following error since the batch size is dropped, and therefore the dimensionality is not matching.  Any clues ? 

`
[192, 16, 16] doesn't match the broadcast shape [48, 192, 16 ,16]
`

Also causal_attention_mask is not used at all in sdpa; don't know how to handle it on below line. 

https://github.com/nileshkokane01/transformers/blob/sdpa_for_OWL_ViT/src/transformers/models/owlvit/modeling_owlvit.py#L484

**Comment by nileshkokane01** [comment]
@younesbelkada  ,

I sought of tried to fix the dimensionality mismatch for batch size , but couldn't figure out. Any clue ? 
```python 
RuntimeError: output with shape 
[192, 16, 16] doesn't match the broadcast shape [48, 192, 16, 16]
```
with these 11 test seems to fail.

**Comment by younesbelkada** [comment]
Hi @nileshkokane01 
Thanks for getting back ! For that I need to deep dive into your branch and try to fix things, I will do that in the next days 🙏

**Pull request closed without merging**